### PR TITLE
Implement time:* 4.0.0 commands

### LINKED ISSUE #625: time:u IStaticService: 400 (GetClockSnapshot) not implemented
IStaticService GetClockSnapshot is not implemented.

## Service description:
[Switchbrew](https://switchbrew.org/wiki/PCV_services#time:u.2C_time:a.2C_time:s) | [Reswitched](https://reswitched.github.io/SwIPC/ifaces.html#nn::timesrv::detail::service::IStaticService(400))

`GetClockSnapshot(u8) -> buffer<nn::time::sf::ClockSnapshot, 0x1a>`

## Required by:
- https://github.com/Ryujinx/Ryujinx-Games-List/issues/136
- https://github.com/Ryujinx/Ryujinx-Games-List/issues/113

### PR DESCRIPTION
**Pull request by Thog**
Round 4~

Close #625

### CODE REVIEW

File: ChocolArm64/Memory/MemoryManager.cs [hunk 1]
```diff
@@ -759,6 +759,19 @@ public void ReadBytes(long position, byte[] data, int startIndex, int size)
             }
         }
 
+        public unsafe T ReadStruct<T>(long position)
```
Comment (gdkchan) [inline ChocolArm64/Memory/MemoryManager.cs]: I don't think this should be added to the CPU memory manager. My current plan is reducing the amount of special methods there (including MemoryManagerHelper).

File: Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs [hunk 2]
```diff
@@ -19,9 +19,14 @@ public long ToSeconds()
             return NanoSeconds / 1000000000;
         }
 
+        public static TimeSpanType FromSeconds(long seconds)
+        {
+            return new TimeSpanType(seconds * 1000000000);
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs]: I'd add a constant for `1000000000`. it avoids having to write this same constant multiple times (which is error prone, for example one 0 missing from the number is hard to notice) and also better documents what this number actually is. Something like:
```cs
private const long NanoSecondsPerSecond = 1000000000;
```

File: Ryujinx.HLE/HOS/Services/Time/Clock/EphemeralNetworkSystemClockCore.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+﻿namespace Ryujinx.HLE.HOS.Services.Time.Clock
+{
+    class EphemeralNetworkSystemClockCore : SystemClockCore
+    {
+        private static EphemeralNetworkSystemClockCore instance;
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Time/Clock/EphemeralNetworkSystemClockCore.cs]: I believe private static fields should be prefixed with a underscore according the C# naming conventions.

File: Ryujinx.HLE/HOS/Services/Time/Clock/StandardLocalSystemClockCore.cs [hunk 4]
```diff
@@ -1,12 +1,7 @@
...
-
-namespace Ryujinx.HLE.HOS.Services.Time.Clock
+﻿namespace Ryujinx.HLE.HOS.Services.Time.Clock
 {
     class StandardLocalSystemClockCore : SystemClockCore
     {
-        private SteadyClockCore    _steadyClockCore;
-        private SystemClockContext _context;
-
         private static StandardLocalSystemClockCore instance;
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Time/Clock/StandardLocalSystemClockCore.cs]: Same as above (and there are probably other instances).

File: Ryujinx.HLE/HOS/Services/Time/Clock/StandardSteadyClockCore.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+{
+    class StandardSteadyClockCore : SteadyClockCore
+    {
+        private long         _setupValue;
+        private ResultCode   _setupResultCode;
+        private bool         _isRtcResetDetected;
+        private TimeSpanType _testOffset;
+        private TimeSpanType _internalOffset;
+
+        private static StandardSteadyClockCore instance;
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Time/Clock/StandardSteadyClockCore.cs]: Same.

File: Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs [hunk 6]
```diff
@@ -59,4 +64,42 @@ struct SystemClockContext
         public long                 Offset;
         public SteadyClockTimePoint SteadyTimePoint;
     }
+
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs]: Remove extra empty line

File: Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs [hunk 7]
```diff
@@ -59,4 +64,42 @@ struct SystemClockContext
...
+        public SteadyClockTimePoint   SteadyClockTimePoint;
+
+        [MarshalAs(UnmanagedType.ByValArray, SizeConst = 0x24)]
+        public char[] LocationName;
+
+        [MarshalAs(UnmanagedType.I1)]
+        public bool   IsAutomaticCorrectionEnabled;
+        public byte   Type;
+        public ushort Unknown;
+
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs]: Remove extra empty line

File: Ryujinx.HLE/HOS/Services/Time/Clock/SystemClockCore.cs [hunk 8]
```diff
@@ -4,11 +4,35 @@ namespace Ryujinx.HLE.HOS.Services.Time.Clock
 {
     abstract class SystemClockCore
     {
-        public abstract SteadyClockCore GetSteadyClockCore();
+        private SteadyClockCore          _steadyClockCore;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/Clock/SystemClockCore.cs]: Remove extra spaces and variables names

File: Ryujinx.HLE/HOS/Services/Time/Clock/TickBasedSteadyClockCore.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+            }
+        }
+
+        private TickBasedSteadyClockCore() {}
+
+        public override SteadyClockTimePoint GetTimePoint(KThread thread)
+        {
+            SteadyClockTimePoint result = new SteadyClockTimePoint
+            {
+                TimePoint = 0,
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/Clock/TickBasedSteadyClockCore.cs]: Align assign

File: Ryujinx.HLE/HOS/Services/Time/IStaticService.cs [hunk 10]
```diff
@@ -116,16 +128,183 @@ public ResultCode IsStandardNetworkSystemClockAccuracySufficient(ServiceCtx cont
...
 
         [Command(300)] // 4.0.0+
-        // CalculateMonotonicSystemClockBaseTimePoint(nn::time::SystemClockContext) -> u64
+        // CalculateMonotonicSystemClockBaseTimePoint(nn::time::SystemClockContext) -> s64
         public ResultCode CalculateMonotonicSystemClockBaseTimePoint(ServiceCtx context)
         {
-            // TODO: reimplement this
-            long timeOffset              = (long)(DateTime.UtcNow - StartupDate).TotalSeconds;
-            long systemClockContextEpoch = context.RequestData.ReadInt64();
+            SystemClockContext otherContext = context.RequestData.ReadStruct<SystemClockContext>();
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/IStaticService.cs]: Remove empty lines, align casts and variables names

File: Ryujinx.HLE/HOS/Services/Time/IStaticService.cs [hunk 11]
```diff
@@ -116,16 +128,183 @@ public ResultCode IsStandardNetworkSystemClockAccuracySufficient(ServiceCtx cont
...
+
+            SteadyClockTimePoint currentTimePoint = StandardSteadyClockCore.Instance.GetCurrentTimePoint(context.Thread);
+
+            ResultCode result = ResultCode.TimeMismatch;
+
+            if (currentTimePoint.ClockSourceId == otherContext.SteadyTimePoint.ClockSourceId)
+            {
+                TimeSpanType ticksTimeSpan = TimeSpanType.FromTicks(context.Thread.Context.ThreadState.CntpctEl0, context.Thread.Context.ThreadState.CntfrqEl0);
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/IStaticService.cs]: Same as previous comment

File: Ryujinx.HLE/HOS/Services/Time/IStaticService.cs [hunk 12]
```diff
@@ -116,16 +128,183 @@ public ResultCode IsStandardNetworkSystemClockAccuracySufficient(ServiceCtx cont
...
+
+            return result;
+        }
+
+        [Command(401)] // 4.0.0+
+        // GetClockSnapshotFromSystemClockContext(u8, nn::time::SystemClockContext, nn::time::SystemClockContext) -> buffer<nn::time::sf::ClockSnapshot, 0x1a>
+        public ResultCode GetClockSnapshotFromSystemClockContext(ServiceCtx context)
+        {
+            byte type = context.RequestData.ReadByte();
+            context.RequestData.BaseStream.Position += 7;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/IStaticService.cs]: Add a new line before

File: Ryujinx.HLE/HOS/Services/Time/IStaticService.cs [hunk 13]
```diff
@@ -116,16 +128,183 @@ public ResultCode IsStandardNetworkSystemClockAccuracySufficient(ServiceCtx cont
...
+
+            return resultCode;
+        }
+
+        private ResultCode GetClockSnapshotFromSystemClockContextInternal(KThread thread, SystemClockContext userContext, SystemClockContext networkContext, byte type, out ClockSnapshot clockSnapshot)
+        {
+            clockSnapshot = new ClockSnapshot();
+
+            SteadyClockCore steadyClockCore = StandardSteadyClockCore.Instance;
+
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/IStaticService.cs]: Remove empty line and then align variables as usual

**Comment by Thog** [comment]
Done

**Pull request merged**